whylog/constraints/__init__.py

```diff
@@ -2,42 +2,184 @@
-        pass
+    """
+    Time delta between two dates must be greater than 'max_delta' and lower than 'min_delta'
+    """
+
+    TYPE = ConstraintType.TIME_DELTA
+
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = 2
+
+    MIN_DELTA = 'min_delta'
+    MAX_DELTA = 'max_delta'
+
+    def __init__(self, param_dict, groups):
+        """
+        :param groups: First element of groups is a group with earlier date.
+
+        I.e:
+        TimeConstraint(
+            {'min_delta': 12, 'max_delta' 30},
+            [(0, 1), (2, 1)]
+        )
+        """
+        super(TimeConstraint, self).__init__(param_dict, groups)
+        self._check_params()
 
+    def _check_params(self):
+        param_names = self.params.keys()
+        if self.MIN_DELTA not in param_names and self.MAX_DELTA not in param_names:
+            raise ConstructorParamsError(self.TYPE, self.get_param_names(), param_names)
+        if not len(self.groups) == 2:
+            raise ConstructorGroupsError(
+                self.TYPE, len(self.groups), self.MIN_GROUPS_COUNT, self.MAX_GROUPS_COUNT
+            )
 
-class IdenticalIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    @classmethod
+    def get_param_names(cls):
+        return [cls.MIN_DELTA, cls.MAX_DELTA]
 
+    @classmethod
+    def verify(cls, param_dict, group_contents):
+        #TODO
+        raise NotImplementedError
 
-class AnyValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
 
+class IdenticalConstraint(AbstractConstraint):
+    """
+    Contents of groups must be identical.
+    """
 
-class DifferentValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    TYPE = ConstraintType.IDENTICAL
 
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = None
 
-class ConstIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    def __init__(self, param_dict, groups):
+        """
+        I.e:
+        IdenticalConstraint(
+            dict(),
+            [(1, 2), (2, 4)]
+        )
+        """
+        super(IdenticalConstraint, self).__init__(param_dict, groups)
 
+    def _check_params(self):
+        if len(self.groups) < self.MIN_GROUPS_COUNT:
+            raise ConstructorGroupsError(
+                self.TYPE, len(self.groups), self.MIN_GROUPS_COUNT, self.MAX_GROUPS_COUNT
+            )
 
-class ValueDeltaIntervals(AbstractConstraint):
-    def __init__(self, interval_lower, interval_greater, min_delta=None, max_delta=None):
+    @classmethod
+    def get_param_names(cls):
+        return []
+
+    @classmethod
+    def verify(cls, param_dict, group_contents):
         """
-        Sets Minimum and maximum difference between values of params (if values are numbers).
+        I.e:
+        - verify({}, ['comp1', 'comp1', 'comp1']) returns True
+        - verify({}, ['comp1', 'hello', 'comp1']) returns False
         """
-        pass
+        if len(group_contents) < 2:
```

Comment:
Well, this is classmethod so it is possible group_contents to have any length. I wasn't sure what to return. I choose false, because verify method should not be called in such cases at all - it is suspicious situation.